sprinkle/high-level/keras.py

```diff
@@ -0,0 +1,73 @@
+    import numpy as np
+    from tensorflow import keras
+    from tensorflow.keras import layers
+
++   import determined as det
++   ctx = det.keras.init()  # also initialized random seeds
++   hparams = ctx.get_hparams()
+
+    # Model / data parameters
+    num_classes = 10
+    input_shape = (28, 28, 1)
+
+    # the data, split between train and test sets
+    (x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
+
+    # Scale images to the [0, 1] range
+    x_train = x_train.astype("float32") / 255
+    x_test = x_test.astype("float32") / 255
+    # Make sure images have shape (28, 28, 1)
+    x_train = np.expand_dims(x_train, -1)
+    x_test = np.expand_dims(x_test, -1)
+    print("x_train shape:", x_train.shape)
+    print(x_train.shape[0], "train samples")
+    print(x_test.shape[0], "test samples")
+
+    # convert class vectors to binary class matrices
+    y_train = keras.utils.to_categorical(y_train, num_classes)
+    y_test = keras.utils.to_categorical(y_test, num_classes)
+
+    model = keras.Sequential(
+        [
+            keras.Input(shape=input_shape),
+            layers.Conv2D(32, kernel_size=(3, 3), activation="relu"),
+            layers.MaxPooling2D(pool_size=(2, 2)),
+            layers.Conv2D(64, kernel_size=(3, 3), activation="relu"),
+            layers.MaxPooling2D(pool_size=(2, 2)),
+            layers.Flatten(),
+            layers.Dropout(0.5),
+            layers.Dense(num_classes, activation="softmax"),
+        ]
+    )
+
+    model.summary()
+
+    batch_size = 128
+    epochs = 15
+
++   # TODO: wrap dataset somewhere
++   optimizer = ctx.keras.wrap_optimizer(...)
+    model.compile(loss="categorical_crossentropy", optimizer=optimizer, metrics=["accuracy"])
+
+    batch_size = 128
++   batch_size = ctx.distributed.per_slot_batch_size(batch_size)  # rename shard_batch_size()
++   cb = ctx.make_callback()
+
+    model.fit(
+        x_train,
+        y_train,
+        batch_size=batch_size,
+        epochs=epochs,
+        validation_split=0.1,
++       initial_epoch=ctx.get_initial_epoch(),
++       callbacks=[cb],
```

Comment:
It happens in the callback, the same as now.  Only since this API would be epoch-based, it would only happen on epoch ends.

Reply:
How does it know which dataset to use?

Reply:
Hm, I wrote this keras API weeks ago and just threw it into this PR last night.  I think it does need some more work yet.

I think I'd rather have a `context.fit()` and a `context.test()` and a `context.predict()`.  Or subclass the model to achieve the same.

What do you think?

Reply:
I think we should avoid dynamically subclassing anything, which is a bit invasive and confusing, since users can totally subclass the model themselves and try to call fit.

context.fit() and a context.test() and a context.predict() are fine to me.

Reply:
ok, I'll resolve the other comments about this file and I'll find some time to re-write it.

Caveat: my goal for right now is not for us to finalize any of the high-level APIs, but rather to have something we believe that we will like and that we can implement.

The really important thing right now is finalizing the low-level APIs.